core/src/impl/Kokkos_BitOps.hpp

```diff
@@ -57,93 +57,193 @@ namespace Kokkos {
 namespace Impl {
 
 KOKKOS_FORCEINLINE_FUNCTION
-int int_log2(unsigned i) {
+int int_log2_fallback(unsigned i) {
   enum : int { shift = sizeof(unsigned) * CHAR_BIT - 1 };
-#if defined(__CUDA_ARCH__) || defined(__HIP_DEVICE_COMPILE__)
+  int offset = 0;
+  if (i) {
+    for (offset = shift; (i & (1 << offset)) == 0; --offset)
+      ;
+  }
+  return offset;
+}
+
+KOKKOS_IMPL_DEVICE_FUNCTION
+inline int int_log2_device(unsigned i) {
+  enum : int { shift = sizeof(unsigned) * CHAR_BIT - 1 };
+#if defined(KOKKOS_ENABLE_CUDA) || defined(KOKKOS_ENABLE_HIP)
   return shift - __clz(i);
-#elif defined(KOKKOS_COMPILER_INTEL)
+#else
+  return int_log2_fallback(i);
```

Comment:
I believe
```suggestion
#elif defined(KOKKOS_COMPILER_INTEL)
  return _bit_scan_reverse(i);  
#else
  return int_log2_fallback(i);
```
would be needed to not change behavior for for Intel compilers on the device.

Reply:
Done